src/program.cpp

```diff
@@ -89,6 +89,73 @@ program::program(program&&) noexcept = default;
 program& program::operator=(program&&) noexcept = default;
 program::~program() noexcept                    = default;
 
+// copy constructor
+program::program(const program& p) { copy(p); }
+
+// copy assignment operator
+program& program::operator=(const program& p)
+{
+    if(this != &p)
+    {
+        copy(p);
+    }
+
+    return *this;
+}
+
+void program::copy(const program& p)
+{
+    // clean the current program
+    if(!impl)
+    {
+        impl = std::make_unique<program_impl>();
+    }
+    else if(!impl->instructions.empty())
+    {
+        remove_instructions(begin(), end());
+    }
+    impl->ctx = p.impl->ctx;
+
+    std::unordered_map<instruction_ref, instruction_ref> ins_map;
+    for(auto ins : iterator_for(p))
+    {
+        instruction_ref copy_ins{};
+        if(ins->name() == "@literal")
+        {
+            auto l   = ins->get_literal();
+            copy_ins = impl->instructions.insert(impl->instructions.end(), instruction{l});
+        }
+        else if(ins->name() == "@param")
+        {
+            auto&& name = any_cast<builtin::param>(ins->get_operator()).parameter;
+            auto s      = ins->get_shape();
+            copy_ins    = impl->instructions.insert(impl->instructions.end(),
+                                                 {builtin::param{name}, std::move(s), {}});
+        }
+        else if(ins->name() == "@outline")
```

Comment:
Do we need this?

Reply:
I have no idea, I checked the function generic_eval(), which processing @outline instructions.

Reply:
I think we are going to remove `@outline` eventually. We dont need it anymore. I guess it can stay there for now.